### ISSUE TITLE
OOM with only ~300MB memory allocated / requested on GPU

### ISSUE THREAD

**Issue by meta-inf** [google/jax#12575]
### Description

The following code crashes with OOM, even though from the error message it doesn't seem to be using or requesting much memory: on an A40 card with ~40GB of available memory, it crashes when having allocated ~340 MB and requesting ~250 MB of memory.

```python
import jax, flax
from jax import numpy as np
from flax import linen as nn
from tqdm import trange

model = nn.Sequential([nn.Dense(50)])

@partial(jax.vmap, in_axes=(0, None))
def sample_from_prior(rng, inp):
    params = model.init(rng, np.zeros((10, 50)))
    return model.apply(params, inp)

for i in trange(1000):
    rngs = jax.random.split(jax.random.PRNGKey(23), 10000)
    sample_from_prior(rngs, np.ones((4, 50)))
    # this won't help
    # if i % 400 == 0:
    #     jax.lib.xla_bridge.get_backend().defragment()
```

A typical error log looks like this:

```
jax._src.traceback_util.UnfilteredStackTrace: jaxlib.xla_extension.XlaRuntimeError: RESOURCE_EXHAUSTED: Out of memory while trying to allocate 250000256 bytes.
BufferAssignment OOM Debugging.
BufferAssignment stats:
             parameter allocation:    78.1KiB
              constant allocation:         0B
        maybe_live_out allocation:   95.37MiB
     preallocated temp allocation:  238.42MiB
  preallocated temp fragmentation:         0B (0.00%)
                 total allocation:  333.86MiB
              total fragmentation:   47.68MiB (14.28%)
Peak buffers:
        Buffer 1:
                Size: 95.37MiB
                Operator: op_name="jit(_truncated_normal)/jit(main)/jit(clip)/min" source_file="/workspace/home/ziyu/anaconda3/envs/jax-c/lib/python3.9/site-packages/flax/core/scope.py" source_line=775
                XLA Label: fusion
                Shape: f32[10000,50,50]
                ==========================

        Buffer 2:
                Size: 47.68MiB
                Operator: op_name="jit(_truncated_normal)/jit(main)/threefry2x32" source_file="/workspace/home/ziyu/anaconda3/envs/jax-c/lib/python3.9/site-packages/flax/core/scope.py" source_line=775
                XLA Label: custom-call
                Shape: u32[10000,1250]
                ==========================

        Buffer 3:
                Size: 47.68MiB
                Operator: op_name="jit(_truncated_normal)/jit(main)/threefry2x32" source_file="/workspace/home/ziyu/anaconda3/envs/jax-c/lib/python3.9/site-packages/flax/core/scope.py" source_line=775
                XLA Label: custom-call
                Shape: u32[10000,1250]
                ==========================

        Buffer 4:
                Size: 47.68MiB
                Operator: op_name="jit(_truncated_normal)/jit(main)/threefry2x32" source_file="/workspace/home/ziyu/anaconda3/envs/jax-c/lib/python3.9/site-packages/flax/core/scope.py" source_line=775
                XLA Label: fusion
                Shape: u32[10000,1250]
                ==========================

        Buffer 5:
                Size: 47.68MiB
                Operator: op_name="jit(_truncated_normal)/jit(main)/threefry2x32" source_file="/workspace/home/ziyu/anaconda3/envs/jax-c/lib/python3.9/site-packages/flax/core/scope.py" source_line=775
                XLA Label: fusion
                Shape: u32[10000,1250]
                ==========================

        Buffer 6:
                Size: 47.68MiB
                Operator: op_name="jit(_truncated_normal)/jit(main)/threefry2x32" source_file="/workspace/home/ziyu/anaconda3/envs/jax-c/lib/python3.9/site-packages/flax/core/scope.py" source_line=775
                XLA Label: fusion
                Shape: u32[10000,1250]
                ==========================

        Buffer 7:
                Size: 78.1KiB
                Entry Parameter Subshape: u32[10000,2]
                ==========================

        Buffer 8:
                Size: 16B
                Operator: op_name="jit(_truncated_normal)/jit(main)/threefry2x32" source_file="/workspace/home/ziyu/anaconda3/envs/jax-c/lib/python3.9/site-packages/flax/core/scope.py" source_line=775
                XLA Label: fusion
                Shape: (u32[10000,1250], u32[10000,1250])
                ==========================

        Buffer 9:
                Size: 16B
                Operator: op_name="jit(_truncated_normal)/jit(main)/threefry2x32" source_file="/workspace/home/ziyu/anaconda3/envs/jax-c/lib/python3.9/site-packages/flax/core/scope.py" source_line=775
                XLA Label: custom-call
                Shape: (u32[10000,1250], u32[10000,1250])
                ==========================

        Buffer 10:
                Size: 4B
                Entry Parameter Subshape: s32[]
                ==========================

        Buffer 11:
                Size: 4B
                Entry Parameter Subshape: s32[]
                ==========================
```

I've tried setting `XLA_PYTHON_CLIENT_MEM_FRACTION` to a smaller value, or using `XLA_PYTHON_CLIENT_ALLOCATOR=platform`, but neither helps.

I should note that this seems specific to some computation performed by flax, in particular the `Sequential` wrapper: it does not happen if we remove it.  Nonetheless it would be better if we don't have to work around it, and the error looks rather strange, so I hope it could get fixed some time.

### What jax/jaxlib version are you using?

0.3.20, 0.3.20+cuda11.cudnn82

### Which accelerator(s) are you using?

GPU

### Additional system info

Python 3.9, Ubuntu 20.04.2, flax 0.6.0

### NVIDIA GPU info

```
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 510.47.03    Driver Version: 510.47.03    CUDA Version: 11.6     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  NVIDIA A40          On   | 00000000:12:00.0 Off |                    0 |
|  0%   36C    P0    86W / 300W |  41147MiB / 46068MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   1  NVIDIA A40          On   | 00000000:13:00.0 Off |                    0 |
|  0%   27C    P8    32W / 300W |      0MiB / 46068MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   2  NVIDIA A40          On   | 00000000:14:00.0 Off |                    0 |
|  0%   27C    P8    31W / 300W |      0MiB / 46068MiB |      0%      Default |
|                               |                      |                  N/A |

# omitted below; code was using an unoccupied card.
```

**Comment by sudhakarsingh27** [comment]
I wasn't able to repro this. Are you able to consistently repro this error? Also, could you make sure there isn't any other process running on the GPUs?

**Comment by meta-inf** [comment]
I think so.

- I can consistently reproduce this on the A40 server (CUDA 11.2), as well as another Linux server with a GTX 2080 Ti (NVIDIA drivier 510.47.03, CUDA 11.4) and the same python / package versions.  If these help: the first server is a docker instance, but the second is bare metal; my `pip freeze` result is [here](https://pastebin.com/i6db6ddh).
- I checked `nvidia-smi` before running the program, and used `CUDA_DEVICE_ORDER=PCI_BUS_ID` to ensure I use the right card.

I find the crash to happen faster if we use a larger network, e.g., `nn.Sequential([nn.Dense(100), nn.Dense(100)])` crashes at iteration 68, and the output is
```
jax._src.traceback_util.UnfilteredStackTrace: jaxlib.xla_extension.XlaRuntimeError: RESOURCE_EXHAUSTED: Out of memory while trying to allocate 1000000256 bytes.
BufferAssignment OOM Debugging.
BufferAssignment stats:
             parameter allocation:    78.1KiB
              constant allocation:         0B
        maybe_live_out allocation:  381.47MiB
     preallocated temp allocation:  953.67MiB
  preallocated temp fragmentation:         0B (0.00%)
                 total allocation:    1.30GiB
              total fragmentation:  190.73MiB (14.28%)                                                                                                                                        
```

**Comment by sudhakarsingh27** [comment]
Ah I can repro now. I think this code is bound to OOM since you're creating a new NN every iteration (do you intend to do that?)

**Comment by zhangqiaorjc** [comment]
If i `print(i)` in the loop, it consistently OOM at step 146 on my machine and 333.86MiB* 146 = 51.5GiB, so it could be hitting total GPU memory limit.

It would seem that somehow the GPU allocations in each loop iteration are not freed, despite trying to delete things at the end of each loop

```
client = jax.lib.xla_bridge.get_backend()

for i in trange(1000):
    print('step=', i)
    bs = [b.shape for b in client.live_buffers()]
    print('num_live_buffers=', len(bs))
    print('live_buffers_shapes=', bs)
    rngs = jax.random.split(jax.random.PRNGKey(23), 10000)
    out = sample_from_prior(rngs, np.ones((4, 50)))
    out.block_until_ready()
    del out
    del rngs
```

```
step= 0
num_live_buffers= 0
live_buffers_shapes= []
  0%|▏                                                                                                                                                     | 1/1000 [00:02<40:32,  2.43s/it]step= 1
num_live_buffers= 3
live_buffers_shapes= [(50,), (10000, 50, 50), (10000, 2)]
step= 2
num_live_buffers= 6
live_buffers_shapes= [(50,), (10000, 50, 50), (10000, 2), (50,), (10000, 50, 50), (10000, 2)]
step= 3
num_live_buffers= 9
live_buffers_shapes= [(50,), (10000, 50, 50), (10000, 2), (50,), (10000, 50, 50), (10000, 2), (50,), (10000, 50, 50), (10000, 2)]```

What made a difference as @meta-inf also mentioned is to avoid using nn.Sequential

```
#model = nn.Sequential([nn.Dense(50)])
model = nn.Dense(50)
```

Now, no more persistent GPU arrays

```
step= 0
num_live_buffers= 0
live_buffers_shapes= []
  0%|▏                                                                                                                                                     | 1/1000 [00:02<41:07,  2.47s/it]step= 1
num_live_buffers= 0
live_buffers_shapes= []
step= 2
num_live_buffers= 0
live_buffers_shapes= []
step= 3
num_live_buffers= 0
live_buffers_shapes= []
  0%|▍                     ```

I'll need to dig into nn.Sequential and see why it's causing this behavior.

**Comment by zhangqiaorjc** [comment]
If you jit the function, it works.

```
step= 999
num_live_buffers= 2
live_buffers_shapes= [(10000, 4, 50), (10000, 2)]
```

Something is holding onto JAX GPU array references, and it's not clear to me why this simple nn.Sequential implementation would lead to that.

But perhaps @jax.jit + nn.Sequential is a ok workaround while I get Flax dev to look into this some more?

Full code
```
import jax, flax
from jax import numpy as np
from flax import linen as nn
from tqdm import trange
from functools import partial

model = nn.Sequential([nn.Dense(50)])
#model = nn.Dense(50)

@jax.jit
@partial(jax.vmap, in_axes=(0, None))
def sample_from_prior(rng, inp):
    params = model.init(rng, np.zeros((10, 50)))
    out = model.apply(params, inp)
    del params
    return out

client = jax.lib.xla_bridge.get_backend()

import gc

for i in trange(1000):
    print('step=', i)
    bs = [b.shape for b in client.live_buffers()]
    print('num_live_buffers=', len(bs))
    print('live_buffers_shapes=', bs)
    rngs = jax.random.split(jax.random.PRNGKey(23), 10000)
    out = sample_from_prior(rngs, np.ones((4, 50)))
    #gc.collect()
    #out.block_until_ready()
    #del out
    #del rngs
    #if i == 5:
    #    break
    
```

**Comment by meta-inf** [comment]
Thanks @ggdakarsingh27 @zhangqiaorjc for the comments and workaround!  I should have thought of `client.live_buffers()`, but I'm not sure why the `BufferAssignment OOM Debugging` log seems to report different things.

And FWIW, I just found it's not limited to `nn.Sequential`.  This also fails if (and only if) we don't `jit`:
```
class Model(nn.Module):
    @nn.compact
    def __call__(self, x):
        x = nn.Dense(200)(x)
        return nn.Dense(200)(x)

model = Model()
```

**Comment by zhangqiaorjc** [comment]
This bug should be fixed if you update to Flax head. See fix in https://github.com/google/flax/issues/2493

**Comment by mattjj** [comment]
Unfortunately, while the code in the OP runs to 1000 steps for me (on an A100 w/ 40GB), following @meta-inf's thinking in [this comment](https://github.com/google/jax/issues/12575#issuecomment-1262801661), a larger model like `Sequential([Dense(250), Dense(250)])` still OOMs for me around step 20 with analogous BufferAssignment stats:

```python
from functools import partial

import jax, flax
from jax import numpy as np
from flax import linen as nn
from tqdm import trange

model = nn.Sequential([nn.Dense(250), nn.Dense(250)])

@partial(jax.vmap, in_axes=(0, None))
def sample_from_prior(rng, inp):
    params = model.init(rng, np.zeros((10, 50)))
    return model.apply(params, inp)

for i in trange(1000):
    print(i)
    rngs = jax.random.split(jax.random.PRNGKey(23), 10000)
    sample_from_prior(rngs, np.ones((4, 50)))
```

```
2022-10-04 03:17:38.522221: E external/org_tensorflow/tensorflow/compiler/xla/pjrt/pjrt_stream_executor_client.cc:2130] Execution of replica 0 failed: RESOURCE_EXHAUSTED: Out of memory while trying to allocate 6250000256 bytes.
BufferAssignment OOM Debugging.
BufferAssignment stats:
             parameter allocation:    78.1KiB
              constant allocation:         0B
        maybe_live_out allocation:    2.33GiB
     preallocated temp allocation:    5.82GiB
  preallocated temp fragmentation:         0B (0.00%)
                 total allocation:    8.15GiB
              total fragmentation:    1.16GiB (14.29%)
```

```
(base) mattjj@matt-a100-8:~$ nvidia-smi
Tue Oct  4 03:18:22 2022
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 515.65.01    Driver Version: 515.65.01    CUDA Version: 11.7     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  NVIDIA A100-SXM...  Off  | 00000000:00:04.0 Off |                    0 |
| N/A   31C    P0    53W / 400W |      0MiB / 40960MiB |      0%      Default |
|                               |                      |             Disabled |
+-------------------------------+----------------------+----------------------+
```

I think this is analogous because the stats show plenty of free memory, yet the allocation failed at runtime. (Please let me know if I'm misinterpreting these.)

We suspect that the reason the code in the OP now runs is that Python's garbage collection just happens to be triggered with that code. IIUC it has heuristics to trigger based on e.g. the number of new objects it sees. But if we make the sizes bigger, like in the larger model case, the Python gc doesn't trigger before we overflow GPU memory.

IIUC google/flax#2493 fixed one issue, but we still have a cyclic references problem.

A workaround for now is to add `gc.collect()` statements to your code, perhaps every few iterations.

**Comment by mattjj** [comment]
I think Flax's `Sequential` is creating reference cycles. I modified jax/core.py's `maybe_find_leaked_tracers` to look like this:

<img width="598" alt="image" src="https://user-images.githubusercontent.com/1458824/193733621-25eae0a5-03f5-45da-b192-bf83539adce5.png">

Then, running the code with `jax.config.update('jax_check_tracer_leaks', True)`, I got this:

```
y is a  <class 'flax.linen.linear.Dense'> 140478991096608
y.parent is a <class 'flax.linen.combinators.Sequential'> 140478991097568
y.parent.layers is a <class 'tuple'> 140478991096416
y.parent.layers[0] is a <class 'flax.linen.linear.Dense'> 140478991096608
```

So I think the issue is a Flax reference cycle. We're talking with the friendly Flax devs to brainstorm a fix! But in the meantime, just doing some `gc.collect()` would solve things as a workaround.
